### PR TITLE
gh-110481: Implement inter-thread queue for biased reference counting

### PR DESCRIPTION
**Pull request by colesbury**
Biased reference counting maintains two refcount fields in each object: `ob_ref_local` and `ob_ref_shared`. The true refcount is the sum of these two fields. In some cases, when refcounting operations are split across threads,
the ob_ref_shared field can be negative (although the total refcount must be at least zero). In this case, the thread that decremented the refcount requests that the owning thread give up ownership and merge the refcount fields.

* Issue: gh-110481

### CODE REVIEW

**Comment by corona10** [comment]
I will take a look by tomorrow :)

File: Include/internal/pycore_object_stack.h
Comment (ericsnowcurrently) [inline Include/internal/pycore_object_stack.h]: Why is this here?

File: Python/brc.c
Comment (ericsnowcurrently) [inline Python/brc.c:32]: This implies a flat distribution of thread IDs.  How confident are we that that's the case?  How much does it matter?

File: Python/brc.c
Comment (ericsnowcurrently) [inline Python/brc.c:145]: Hmm, I missed that we added another "thread ID" value, to go along with `PyThreadState.thread_id` (and `.native_thread_id`).  What's the advantage of this thread ID over the system one?  (This is mostly just me being curious.)

File: Python/brc.c
Comment (ericsnowcurrently) [inline Python/brc.c]: If I understand correctly, this means that all other thread states in the bucket would block on BRC operations until the active BRC state merge finishes.  Is that right?  I'm guessing that making them block past the `llist_remove()` is unnecessary but makes the code here simpler.  Or have I misunderstood?

File: Python/brc.c
Comment (ericsnowcurrently) [inline Python/brc.c:64]: This smells like a short-circuit to avoid doing the rest of this function unnecessarily.  Is that right?  (There is still a benign race between here and later where `ob_tid` could be cleared out by the owning thread, where at worst we do some unnecessary work.)

File: Python/brc.c
Comment (ericsnowcurrently) [inline Python/brc.c:70]: ...or the owning `tid` applies to a thread state in a different interpreter.  Are we still okay here in that case?

The one-to-many relationship between `ob_tid` and thread states makes me a little uneasy.  (The same goes for `brc.tid`, I suppose.)  From what I understand, `ob_ref_local` relates to the thread where `new_reference()` was last called on the object.  It is *not* tied to a specific thread state.  (This impacts functions like `_Py_ExplicitMergeRefcount()`.)

That's relevant to this PR since the BRC code here *is* tied to a specific thread state.  I'm just unclear if that's an actual problem or not.

Relatedly, something we've considered is storing a reference to the owning `PyInterpreterState` on each `PyTypeObject`.  Then you can get the owning interpreter from `ob_type`.  (It would be a little tricky for builtin static types though.)  Perhaps that would help here (if needed)?

File: Python/brc.c
Comment (ericsnowcurrently) [inline Python/brc.c:97]: It certainly would be nice to have a per-thread eval breaker, for situations like this.

**Comment by ericsnowcurrently** [comment]
FTR, biased refcounting was implemented in gh-110764.

File: Python/gc_free_threading.c
Comment (ericsnowcurrently) [inline Python/gc_free_threading.c:319]: I'm not sure if it matters in places like this, but do note that one interpreter can hold a reference to an object owned by another interpreter.  Furthermore, the two interpreters may have thread states with the same thread ID and `ob_tid` may be set to that thread ID.  I'm getting the feeling that there may be some trickiness with this.

File: Python/gc_free_threading.c
Comment (ericsnowcurrently) [inline Python/gc_free_threading.c:314]: We really should have a per-interpreter lock for this, rather than blocking all interpreters.  Something like `PyInterpreterState.threads.mutex` (a `PyMutex`).  Is there an issue open for that already?

File: Lib/test/test_code.py
Comment (ericsnowcurrently) [inline Lib/test/test_code.py:868]: If I understand right, `gc.collect()` is how a user can make sure all objects that need to be finalized/deleted have been.  Is that clear in the docs?

File: Include/internal/pycore_brc.h
Comment (ericsnowcurrently) [inline Include/internal/pycore_brc.h:7]: Why not put these after the `#ifndef Py_BUILD_CORE` block?

File: Include/internal/pycore_brc.h
Comment (ericsnowcurrently) [inline Include/internal/pycore_brc.h:20]: Thanks for a nice, brief explanation!

File: Include/internal/pycore_brc.h
Comment (ericsnowcurrently) [inline Include/internal/pycore_brc.h]:
```suggestion
    // Hash table of thread states by thread-id. Thread states within a bucket are
```

File: Include/internal/pycore_brc.h
Comment (ericsnowcurrently) [inline Include/internal/pycore_brc.h]:
```suggestion
    // Mutex protects both the bucket and thread state queues in this bucket.
```

File: Python/brc.c
Comment (ericsnowcurrently) [inline Python/brc.c:81]: Can we be more efficient when `tstate == _PyThreadState_GET()`?

**Review by ericsnowcurrently — CHANGES_REQUESTED** [review]
The approach generally makes sense.  I'm pretty sure I was able to follow along.  There were a few places where I'm concerned there may be some extra complexity with multiple interpreters but they may be non-issues.  I've left a handful of other minor comments.

**Comment by ericsnowcurrently** [comment]
FWIW, every once in a while I'm reminded about the extra complexity that no-gil brings to GC and refcounting.  I still worry that the extra complexity will have a meaningful long-term impact on maintainability for CPython and community extension modules.  Is there anything we can do to help contributors/maintainers understand how it all works under no-gil and especially how to track down leaks/crashes without breaking their brains?  Is there any extra tooling that would help?

File: Python/brc.c
Comment (corona10) [reply Python/brc.c:32]: Yeah, I have the same question as Eric with this assumption.

1. How many unique tids in a stressful situation are observed.
2. There is no possibility of skewed bucket issue?
3. If there is a possibility of a skewed bucket issue, can we use tree traversing based on tid?
(But.. yeah in this case we need balanced tree :( )

File: Include/internal/pycore_brc.h
Comment (colesbury) [reply Include/internal/pycore_brc.h:7]: I'm trying to avoid the `#include` inside the `extern "C" {`. That can lead to compilation problems with system headers (or headers that transitively include system headers).

File: Include/internal/pycore_object_stack.h
Comment (colesbury) [reply Include/internal/pycore_object_stack.h]: We use the `_PyFreeListState *` on the line below so we need to either "forward declare" it or include `pycore_freelist.h`. I can switch to an `#include` if you prefer.

File: Lib/test/test_code.py
Comment (colesbury) [reply Lib/test/test_code.py:868]: In the general sense, that's long been true, which is why we have `gc_collect()` sprinkled throughout tests.

In terms of docs, I think we will want to call out the free-threading specific differences both in the "What's New" section and on a free-threading docs page. I want to do both of those things later when the integration is farther along.

File: Python/brc.c
Comment (colesbury) [reply Python/brc.c:32]: The thread ids don't need to be uniformly distributed, they just can't be correlated with the number of buckets, which is why we chose a prime size instead of a power-of-two (less likely to share common denominators).

This is essentially the same as the parking_lot "waiters" hash table, but per-interpreter instead of truly global. It has the same limitations: if we have many thousands of threads, the chaining will be slow. We'll want to address that eventually in both places, either by using a balanced tree or resizing the table. (I think a balanced tree would probably be simpler).

I don't think that's necessary yet; we'll have a lot more urgent scaling issues to address before then.

File: Python/brc.c
Comment (colesbury) [reply Python/brc.c:64]: You are correct that there is still a benign race where the object gets queued while it's being merged. The owning thread will call `_Py_ExplicitMergeRefcount()` twice on the same object, but that's [fine](https://github.com/python/cpython/blob/ee66c333493105e014678be118850e138e3c62a8/Objects/object.c#L402-L404).

This avoids a slightly different race that's more problematic. Without this check, the decref'ing thread would interpreter the zero `ob_tid` as if the owning thread had exited, because it would not find any thread with `ob_tid` of zero. That could lead to both the decref'ing thread and the owning thread calling `_Py_ExplicitMergeRefcount()` on the same object concurrently, which is not safe.

There are other ways we could have addressed this problem. We could have instead ensured that owning thread zeros `ob_tid` only *after* it finishes merging the refcount, but I think that the check here is more robust.

File: Python/brc.c
Comment (colesbury) [reply Python/brc.c:81]: This code path is only executed for `tstate != _PyThreadState_GET()`. We don't call this function for objects "owned" by the current thread.

File: Python/brc.c
Comment (colesbury) [reply Python/brc.c:97]: Yeah, per-thread eval breaker will be important for this. Brett Simmers is working on that.

File: Python/brc.c
Comment (colesbury) [reply Python/brc.c:145]: `_PyThread_Id()` is faster than retrieving any of the other thread ID values from `PyThreadState`. That's important because it's used in almost every refcount operation.

File: Python/brc.c
Comment (colesbury) [reply Python/brc.c]: Yes, that's correct. Unfortunately, there's a bug and the code needs to be changed.

FWIW, the `_PyObjectStack_Merge()` calls in this file are very cheap. The `local_objects_to_merge` stack is almost always empty -- it only has items when a finalizer reentrantly calls back into the queue processing code, which is rare.

File: Python/gc_free_threading.c
Comment (colesbury) [reply Python/gc_free_threading.c:314]: I'm not aware of one. If there is one, I suspect you are the issue author :-)

File: Python/gc_free_threading.c
Comment (colesbury) [reply Python/gc_free_threading.c:319]: Having thread states in different interpreters with the same thread id is fine, and we have tests that exercise that behavior. The mapping is effectively `(PyInterpreterState, ob_tid) -> PyThreadState`.

File: Python/brc.c
Comment (colesbury) [reply Python/brc.c:70]: > ...or the owning tid applies to a thread state in a different interpreter. Are we still okay here in that case?

This doesn't make that case any better or worse. The thread might be decref'd in a different interpreter than it was created in, but it's the same interpreter as the `Py_DECREF()` call.

Allocating an object in one interpreter and deallocating in another is problematic both in the default build and in the free-threaded build. We have some tests like `SinglephaseInitTests.test_basic_multiple_interpreters_deleted_no_reset`, but they only test really basic objects. Once you have non-trivial Python destructors you will encounter segfaults.

> Relatedly, something we've considered is storing a reference to the owning PyInterpreterState on each PyTypeObject.

That might help debugging, but I don't think it's necessary for this PR.

**Review by colesbury — COMMENTED** [review]
> Is there anything we can do to help contributors/maintainers understand how it all works under no-gil and especially how to track down leaks/crashes without breaking their brains? Is there any extra tooling that would help?

There are some updates to the devguide (for GC) under review. We'll want to do more along with user facing docs, but I think these are best addressed holistically once the integration is farther along.

On tooling, we already have a nogil refleaks buildbot. We'll have a thread sanitizer buildbot after we get to the point where we can actually disable the GIL.

File: Include/internal/pycore_object_stack.h
Comment (ericsnowcurrently) [reply Include/internal/pycore_object_stack.h]: I suppose I'm fine either way, but I'd lean toward pulling it in with the include.  I tend to favor a single-source-of-truth approach.  In the code base I think we generally go the include route (but that's just my 2-second mental review of experience 😄).

File: Python/brc.c
Comment (ericsnowcurrently) [reply Python/brc.c:70]: We definitely shouldn't decref an object in one interpreter that belongs to another interpreter.  It's just that in the owning interpreter we may incref the object before sending it to another interpreter, just to make sure it stays alive.  In that scenario, the owning interpreter must still be where the decref happens.  For example, the other interpreter can release its reference by queueing up the decref as a "pending call" on the owning interpreter.

Between what you said and me having thought about it a little more, I think the `ob_tid` one-to-may relationship isn't a problem here.  I don't know if it might impact any other code.

File: Python/brc.c
Comment (ericsnowcurrently) [reply Python/brc.c:97]: That's great to hear.  Is there an issue open for that?

Comment (colesbury) [reply Python/brc.c:97]: Yup, it's here: https://github.com/python/cpython/issues/112175

File: Python/gc_free_threading.c
Comment (ericsnowcurrently) [reply Python/gc_free_threading.c:314]: I've created gh-114940.

File: Include/internal/pycore_brc.h
Comment (ericsnowcurrently) [reply Include/internal/pycore_brc.h:7]: It seems like we've been okay in other header files.  Did something change?  Perhaps we simply weren't aware of the problem?  Should we go through our other header files and change to this pattern?

Comment (colesbury) [reply Include/internal/pycore_brc.h:7]: The problem only exists if these header files are compiled as C++ code. It's a bit weird because these are *internal* header files and CPython is in C (not C++), but we've added `extern "C" {` to them so it seems like we intend to support C++ compilation. Sometimes embedders or naughty C-API extensions include the `pycore_xxx` headers (possibly in a C++ file).

I think we should fix the other internal headers, but no one has complained yet. Most likely that's because:

1) It's rare to compile one of the `pycore_xxx.h` headers as C++
2) Most of the `pycore_xxx.h` headers don't include system headers.

Here's an example of the same issue in our public header files, which Victor fixed: https://github.com/python/cpython/issues/110845#issuecomment-1805975034

Comment (ericsnowcurrently) [reply Include/internal/pycore_brc.h:7]: ping @gpshead

I'm sure you have some insight on this. 😄

**Comment by colesbury** [comment]
The "Ubuntu (free-threading)" build crashes due to the problem described in https://github.com/python/cpython/issues/115035. I think that will need to be fixed before this.

File: Include/internal/pycore_brc.h
Comment (colesbury) [reply Include/internal/pycore_brc.h:7]: I filed https://github.com/python/cpython/issues/115105 to track and discuss this

**Comment by colesbury** [comment]
Hi @ericsnowcurrently, I believe I've addressed the comments from your review. Would you please take another look?

**Review by ericsnowcurrently — APPROVED** [review]
LGTM

**Comment by colesbury** [comment]
!buildbot nogil

**Review by corona10 — APPROVED** [review]

**Comment by colesbury** [comment]
There were some issues with refleak tracking. After some failed attempts at whack-a-mole, I addressed the refleak issues in a separate PR (https://github.com/python/cpython/pull/115188), and re-merged main back into this PR. It looks like the buildbots are finally passing 😅

**Pull request merged**
